### DISCUSSION TITLE
@WithSpan annotation doesn't seem to work?

### DISCUSSION THREAD

**Discussion by stevensim226** [open-telemetry/opentelemetry-java#3137] [Q&A]
Hello there,
From what I read in this [documentation](https://opentelemetry.io/docs/java/automatic_instrumentation/#manual-instrumentation) there is a feature to open a a span by just using an annotation.

My code is as follows:
 ```
Span span = tracer.spanBuilder("ping").startSpan();
printTrace();
span.end();
```
with the method `printTrace()` as follows
```
@WithSpan(value = "printTrace")
public void printTrace() {
    System.out.println("Hello World!");
}
```

And from what I am using now (JaegerExporter) it only shows the `ping` span, but the `printTrace` span is nowhere to be found.
I have also added the java agent for automatic instrumentation (even though I don't think this is required for using the annotation `@WithSpan`)

Is there something that I forgot to set?

**Comment by kenfinnigan** [comment]
Is the `opentelemetry-extension-annotations` dependency present?

**Reply by stevensim226** [reply]
Hi @kenfinnigan 
I have just changed the old dependency to the dependency that you mentioned.
But unfortunately it still hasn't exported the `printTrace` span yet, only the `ping` span is exported still to Jaeger

Although I did find an interesting error in the logs as follows
```
[opentelemetry.auto.trace 2021-04-08 19:46:11:566 +0700] [BatchSpanProcessor_WorkerThread-1] WARN io.opentelemetry.exporter.otlp.trace.OtlpGrpcSpanExporter - Failed to export spans. Error message: UNAVAILABLE: io exception
```
From what I guess it seems to be trying to export the `printTrace` span through `OtlpGgrpcSpanExporter` even though I am using `JaegerGrpcSpanExporter`, if my assumption is correct then maybe is there a setting that I missed to set the exporter to be used by `@WithSpan` annotation?

**Reply by kenfinnigan** [reply]
One thing to note is the dependency for `@WithSpan` doesn't include any implementation, it's just an API.

For `@WithSpan` to work there needs to be an implementation to process the annotation with an interceptor, or the like. The documentation seems to provide an example of auto instrumentation to provide an implementation but I'm not certain if there is one at present.

**Reply by stevensim226** [reply]
Hello @kenfinnigan 
I'm currently using manual instrumentation, and I found the docs for `@WithSpan` annotation here https://opentelemetry.io/docs/java/automatic_instrumentation/#configure-withspan-annotation and it seems that I have done it as how it's written

For your information, my tracer instantiation code is like this
```
Tracer provideTracer() {
        JaegerGrpcSpanExporter exporter =
                JaegerGrpcSpanExporter.builder()
                        .setEndpoint("http://localhost:14250")
                        .build();

        // Build resource to set service name manually due to library not yet updated
        AttributesBuilder attributesBuilder = Attributes.builder();
        attributesBuilder.put("service.name", "sample-tracer");
        Resource resource = Resource.create(attributesBuilder.build());

        SdkTracerProvider sdkTracerProvider = SdkTracerProvider.builder()
                .setResource(resource)
                .addSpanProcessor(BatchSpanProcessor.builder(exporter).build())
                .build();

        OpenTelemetry openTelemetry = OpenTelemetrySdk.builder()
                .setTracerProvider(sdkTracerProvider)
                .setPropagators(ContextPropagators.create(W3CTraceContextPropagator.getInstance()))
                .buildAndRegisterGlobal();

        return openTelemetry.getTracer("sample-tracer");
    }
```

And I'm currently using Jaeger Query through the Docker Image `jaegertracing/all-in-one:1.22` w/ these ports opened to localhost
- 14250/tcp localhost:14250
- 14268/tcp localhost:14268
- 16686/tcp localhost:16686
- 5775/udp localhost:5775
- 5778/tcp localhost:5778
- 6831/udp localhost:6831
- 6832/udp localhost:6832
- 9411/tcp localhost:9411

I'm not really good in computer networks but perhaps anything wrong with the `setEndpoint("localhost:14250")` ?

**Reply by kenfinnigan** [reply]
The Javadoc mentions the need for an implementation to process the annotation: https://www.javadoc.io/doc/io.opentelemetry/opentelemetry-contrib-auto-annotations/latest/index.html

Also, you mention you are using manual instrumentation but the `@WithSpan` is documented in the automatic instrumentation docs. It sounds like that's where the mismatch is which is why it's not being exported

**Reply by jkwatson** [reply]
At the moment the only thing that provides an implementation of the annotation is the auto instrumentation agent. 

We have an open feature request to provide alternative ways of using it, but nothing has been implemented yet.

**Reply by stevensim226** [reply]
I see @jkwatson ,
Though I now already understood that it seems the annotation is only usable if automatic instrumentation is enabled and this may not be the appropriate place to ask this question, but what's the purpose of `@WithSpan` if we already enabled automatic instrumentation?

Or CMIIW but is it like the automatic instrumentation's purpose is to solely provide the tracer instance?

**Reply by kenfinnigan** [reply]
The intention is to allow addition span creation for internal methods, as you had done, that automatic instrumentation wouldn't add by default

**Reply by ghost** [reply]
@stevensim226 this is the feature request / issue that @jkwatson mentioned about processing the annotations without autoinstrumentation agent: https://github.com/open-telemetry/opentelemetry-java/issues/3035
